Bluetooth: Mesh: Add Config Client API

haoyanbiao opened the pull request:
The api is used to reset a node (other than a Provisioner) and
remove it from the network

Signed-off-by: YanBiao Hao <haoyanbiao@xiaomi.com>

trond-snekvik commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -1125,6 +1140,39 @@ int bt_mesh_cfg_app_key_add(uint16_t net_idx, uint16_t addr, uint16_t key_net_id
+	};
+	int err;
+
+	err = cli_prepare(NULL, OP_NODE_RESET_STATUS);
```
Should get the status from the response status, here and in reset status callback.

trond-snekvik commented on doc/reference/bluetooth/mesh/shell.rst:
```diff
@@ -134,6 +134,12 @@ General configuration
+``mesh node reset``
+-------------------
+
+        Reset a node (other than a Provisioner) and remove it from the network
```
Please indent with a single tab character

trond-snekvik commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -2607,6 +2607,22 @@ static int cmd_cdb_app_key_del(const struct shell *shell, size_t argc,
+	shell_print(shell, "Node reset complete");
+
+	return 0;
+}
```
Could this be merged with the reset command instead? ie keep the current reset command, but do this inside. It should be possible to reset the local device with the config message, and having two very similar commands is confusing.

trond-snekvik commented on doc/reference/bluetooth/mesh/shell.rst:
```diff
@@ -128,10 +128,11 @@ General configuration
 
-	Reset the mesh node to its initial unprovisioned state.
+	Reset the local mesh node to its initial unprovisioned state or reset a remote node and remove it from the network.
+	* ``value``: Sets local node or remote node.
```
The local/remote decision is made by the `dst` address in all other APIs, I think this should follow the same pattern. 
The command should also be moved into the Configuration client section.

trond-snekvik commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -1125,6 +1150,42 @@ int bt_mesh_cfg_app_key_add(uint16_t net_idx, uint16_t addr, uint16_t key_net_id
+		return 0;
+	}
+
+	return cli_wait();
```
As the status now only depends on whether the response message was received or not, it'll either be `true` or undefined, as we never explicitly set it to `false`. It should either be set to false before sending the message, or removed.

trond-snekvik commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -216,6 +216,30 @@ static void net_key_list(struct bt_mesh_model *model,
 
+struct node_reset_param {
+	bool *status;
+};
```
No need for this structure, really. The raw `bool *` can be used as an `op_param`.

trond-snekvik commented:
In addition to the other comments: Please recheck your sign-off line format again. There should generally only be one, and it should match the commit author email. It should also be preceded by a blank line.

LingaoM commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -428,8 +428,31 @@ static int cmd_uuid(const struct shell *shell, size_t argc, char *argv[])
+		bt_mesh_reset();
+		shell_print(shell, "Local node reset complete");
+	} else {
+		uint8_t reset;
```
Type bool here.

LingaoM reviewed (changes requested):
Please check you sign-off line format, I’m notice that commit sign-off email inconsistent with git author.

Typically use `git commit -s` to input.

For more information, please see [here](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-guidelines) in detail.

LingaoM commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -216,6 +216,26 @@ static void net_key_list(struct bt_mesh_model *model,
+static void node_reset_status(struct bt_mesh_model *model,
+		struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf)
+{
+	struct bool *param = NULL;
```
`struct  bool *param = NULL;` and below `*param->status`

LingaoM replied:
`bool` is a standard type defined by C programs, and other type declarations are not allowed as prefixes.

LingaoM replied:
You have some build error, please click Details in `buildkite/zephyr` below for more information.

Commit:
Bluetooth: Mesh: Add Config Client API

The api is used to reset a node (other than a Provisioner) and
remove it from the network

Signed-off-by: YanBiao Hao <haoyanbiao@xiaomi.com>

trond-snekvik reviewed (changes requested):
The status parameter still needs to be set to `false` inside `bt_mesh_cfg_node_reset`. Other than that, this LGTM.

trond-snekvik commented:
@haoyanbiao if you don't mind, I'll amend the commit to clear the status, so we can get this merged

haoyanbiao commented:
> @haoyanbiao if you don't mind, I'll amend the commit to clear the status, so we can get this merged

There is a lot of work to do recently. sorry about that I didn't check your information in time. Of course, you can modify it directly

trond-snekvik reviewed (approved):
I pushed a change to make sure status is always set now. The rest LGTM :+1:

LingaoM commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -428,8 +428,30 @@ static int cmd_uuid(const struct shell *shell, size_t argc, char *argv[])
+		if (err) {
+			shell_error(shell, "Unable to send"
+					"Remote Node Reset (err %d)", err);
+			return 0;
```
Is a space missing in the `shell_error` ?

LingaoM commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -1125,6 +1146,44 @@ int bt_mesh_cfg_app_key_add(uint16_t net_idx, uint16_t addr, uint16_t key_net_id
 
+int bt_mesh_cfg_node_reset(uint16_t net_idx, uint16_t addr, bool *status)
+{
+	BT_MESH_MODEL_BUF_DEFINE(msg, OP_NODE_RESET, 2);
```
Config Node Reset no need for payloads, of course, the length.
`BT_MESH_MODEL_BUF_DEFINE(msg, OP_NODE_RESET, 0);`

LingaoM commented:
Other LGTM 👍
